MetaMask/metamask-extension
# Add EIP-6963 Provider

## Explanation

Helps to alleviate `window.ethereum` conflicts by supporting an alternative way for Dapps to discover and interact with providers. This PR implements [EIP-6963](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6963.md) which adds an async event based provider discovery standard. 

~~NOTE:  I'm not sure if we strictly need the changes I made that add EIP-6963 support to the test dapp, but we probably want to add an e2e test for EIP-6963 and I'm not sure if we want our tests to rely on an external site like https://eip6963.org/ . Maybe we can test this by running some js in page like we do for testing parts of the json rpc in e2e environment.~~ I've added a spec for this that tests it directly in the browser console instead

Ticket: [mmp-869](https://github.com/MetaMask/MetaMask-planning/issues/869)
Related: https://github.com/MetaMask/providers/pull/263
~~Related: https://github.com/MetaMask/test-dapp/pull/243~~

# Screenshot
<img width="470" alt="Screenshot 2023-09-27 at 4 41 13 PM" src="https://github.com/MetaMask/metamask-extension/assets/918701/88a0e334-fed9-45e8-80a2-329cb4c94ded">

## Manual Testing Steps

* Install and enable a few other browser extension wallets
* Visit https://eip6963.org/
* MetaMask should show up in the list of providers and play nicely with all of them

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
